src/patchelf.cc

```diff
@@ -1274,26 +1312,109 @@ void ElfFile<ElfFileParamNames>::rewriteHeaders(Elf_Addr phdrAddress)
             unsigned int shndx = rdi(sym->st_shndx);
-            if (shndx != SHN_UNDEF && shndx < SHN_LORESERVE) {
-                if (shndx >= sectionsByOldIndex.size()) {
-                    fprintf(stderr, "warning: entry %d in symbol table refers to a non-existent section, skipping\n", shndx);
-                    continue;
+
+            if (shndx == SHN_UNDEF || shndx >= SHN_LORESERVE) {
+                continue;
+            }
+
+            if (shndx >= sectionsByOldIndex.size()) {
+                warn(
+                    "entry %d refers to a non-existent section (%d), skipping\n",
+                    entry,
+                    shndx
+                );
+
+                continue;
+            }
+
+            const std::string & section = sectionsByOldIndex.at(shndx);
+            assert(!section.empty());
+
+            auto newIndex = getSectionIndex(section);
+            auto newValue = rdi(sym->st_value);
+
+            if (ELF32_ST_TYPE(rdi(sym->st_info)) == STT_SECTION) {
+                newValue = rdi(shdrs.at(newIndex).sh_addr);
+            } else {
+                auto rs = replacedSections.find(section);
+
+                if (rs == replacedSections.end()) {
+                    // Entry refers to a section that didn't got moved around
+                    // (we still might have to update shndx, though)
+                } else {
+                    if (rdi(sym->st_value) < rs->second.prev_addr) {
+                        warn("entry %d refers to an address outside the section, skipping\n", entry);
+                        continue;
+                    } else {
+                        newValue = rdi(sym->st_value) - rs->second.prev_addr + rdi(shdrs.at(newIndex).sh_addr);
+                    }
+                }
+            }
+
+            auto oldIndex = rdi(sym->st_shndx);
+            auto oldValue = rdi(sym->st_value);
+
+            if (newIndex == oldIndex && newValue == oldValue) {
+                // No point in spamming logs if nothing got changed
+                continue;
+            }
+
+            debug(
+                "rewriting entry %d: shndx=%d->%d (%s), value=%d->%d\n",
+                entry,
+                (size_t) oldIndex,
+                (size_t) newIndex,
+                section.c_str(),
+                (size_t) oldValue,
+                (size_t) newValue
+            );
+
+            wri(sym->st_shndx, newIndex);
+            wri(sym->st_value, newValue);
+        }
+    }
+
+    /* Rewrite the .rela.dyn section */
+    for (unsigned int i = 1; i < rdi(hdr()->e_shnum); ++i) {
+        auto &shdr = shdrs.at(i);
+
+        if (rdi(shdr.sh_type) != SHT_RELA) {
+            continue;
+        }
+
+        debug("rewriting relocations (section %d)\n", i);
+
+        unsigned int r_idx = 0;
+
+        for (auto& r: getSectionSpan<Elf64_Rela>(shdr)) {
+            if (ELF64_R_TYPE(r.r_info) == R_X86_64_RELATIVE) {
```

Comment:
aarch64 relatives (and possibly others) have the same format, maybe we could support them as well?